frontend/packages/ux-editor/src/components/config/editModal/EditOptions/OptionTabs/EditOptionChoice/EditOptionList/EditOptionList.tsx

```diff
@@ -0,0 +1,159 @@
+import { useTranslation } from 'react-i18next';
+import { StudioDropdownMenu, StudioFileUploader, StudioSpinner } from '@studio/components';
+import { BookIcon } from '@studio/icons';
+import { useStudioEnvironmentParams } from 'app-shared/hooks/useStudioEnvironmentParams';
+import { FileNameUtils } from '@studio/pure-functions';
+import { findFileNameError } from './utils/findFileNameError';
+import type { FileNameError } from './utils/findFileNameError';
+import type { AxiosError } from 'axios';
+import type { ApiError } from 'app-shared/types/api/ApiError';
+import { toast } from 'react-toastify';
+import classes from './EditOptionList.module.css';
+
+type EditOptionListProps<T extends SelectionComponentType> = {
+  setChosenOption: (value: boolean) => void;
+} & Pick<IGenericEditComponent<T>, 'component' | 'handleComponentChange'>;
+
+export function EditOptionList<T extends SelectionComponentType>({
+  setChosenOption,
+  component,
+  handleComponentChange,
+}: EditOptionListProps<T>) {
+  const { t } = useTranslation();
+  const { org, app } = useStudioEnvironmentParams();
+  const { data: optionListIds } = useOptionListIdsQuery(org, app);
+  const { mutate: uploadOptionList } = useAddOptionListMutation(org, app, {
+    hideDefaultError: (error: AxiosError<ApiError>) => !error.response.data.errorCode,
+  });
+
+  const handleOptionsIdChange = (optionsId: string) => {
+    if (component.options) {
+      delete component.options;
+    }
+
+    handleComponentChange({
+      ...component,
+      optionsId,
+    });
+
+    setChosenOption(true);
+  };
+
+  const onSubmit = (file: File) => {
+    const fileNameError = findFileNameError(optionListIds, file.name);
+    if (fileNameError) {
+      handleInvalidFileName(fileNameError);
+    } else {
+      handleUpload(file);
+    }
+  };
+
+  const handleUpload = (file: File) => {
+    uploadOptionList(file, {
+      onSuccess: () => {
+        handleOptionsIdChange(FileNameUtils.removeExtension(file.name));
+        toast.success(t('ux_editor.modal_properties_code_list_upload_success'));
+      },
+      onError: (error: AxiosError<ApiError>) => {
+        if (!error.response?.data?.errorCode) {
+          toast.error(`${t('ux_editor.modal_properties_code_list_upload_generic_error')}`);
+        }
+      },
+    });
+  };
+
+  const handleInvalidFileName = (fileNameError: FileNameError) => {
+    switch (fileNameError) {
+      case 'invalidFileName':
+        return toast.error(t('ux_editor.modal_properties_code_list_filename_error'));
+      case 'fileExists':
+        return toast.error(t('ux_editor.modal_properties_code_list_upload_duplicate_error'));
+    }
+  };
+
+  return (
+    <>
+      <OptionListSelector handleOptionsIdChange={handleOptionsIdChange} />
+      <StudioFileUploader
+        accept='.json'
+        variant={'tertiary'}
+        uploaderButtonText={t('ux_editor.options.upload_title')}
+        onSubmit={onSubmit}
+      />
+    </>
+  );
+}
```

Comment:
Can the selector and the file uploader be separated into two different components?

Suggestion:
`handleOptionsIdChange` could be moved into `OptionListSelector`, and a new component `OptionListUploader` could wrap around `StudioFileUploder` and its handler functions. This would improve separation of concerns.

Then we wouldn't need `EditOptionList` anymore, and `OptionListSelector` and `OptionListUploader` could be called directly from `EditOptionChoice`.

`EditOptionChoice` could then return this:
```tsx
{chosenOption ? (
        <DisplayChosenOption
          setChosenOption={setChosenOption}
          component={component}
          handleComponentChange={handleComponentChange}
        />
      ) : (
        <div className={classes.optionButtons}>
          <EditManualOptionsWithEditor
            setChosenOption={setChosenOption}
            component={component}
            handleComponentChange={handleComponentChange}
          />
          <OptionListSelector
            setChosenOption={setChosenOption}
            component={component}
            handleComponentChange={handleComponentChange}
          />
          <OptionListUploader
            setChosenOption={setChosenOption}
            component={component}
            handleComponentChange={handleComponentChange}
          />
        </div>
```

Reply:
Nice work!